flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesClusterDescriptor.java

```diff
@@ -0,0 +1,186 @@
+
+	public KubernetesClusterDescriptor(Configuration flinkConfig, FlinkKubeClient client) {
+		this.flinkConfig = flinkConfig;
+		this.client = client;
+		this.clusterId = flinkConfig.getString(KubernetesConfigOptions.CLUSTER_ID);
+		Preconditions.checkNotNull(clusterId, "ClusterId must be specified!");
+	}
+
+	@Override
+	public String getClusterDescription() {
+		return CLUSTER_DESCRIPTION;
+	}
+
+	private ClusterClient<String> createClusterClient(String clusterId) throws Exception {
+
+		Configuration configuration = new Configuration(flinkConfig);
+
+		final Endpoint restEndpoint = this.client.getRestEndpoint(clusterId);
+
+		if (restEndpoint != null) {
+			configuration.setString(RestOptions.ADDRESS, restEndpoint.getAddress());
+			configuration.setInteger(RestOptions.PORT, restEndpoint.getPort());
+		} else {
+			throw new ClusterRetrieveException("Could not get the rest endpoint of " + clusterId);
+		}
+
+		return new RestClusterClient<>(configuration, clusterId);
+	}
+
+	@Override
+	public ClusterClient<String> retrieve(String clusterId) throws ClusterRetrieveException {
+		try {
+			final ClusterClient<String> retrievedClient = createClusterClient(clusterId);
+			LOG.info("Retrieve flink cluster {} successfully, JobManager Web Interface : {}", clusterId,
+				retrievedClient.getWebInterfaceURL());
+			return retrievedClient;
+		} catch (Exception e) {
+			this.client.handleException(e);
+			throw new ClusterRetrieveException("Could not create the RestClusterClient.", e);
+		}
+	}
+
+	@Override
+	public ClusterClient<String> deploySessionCluster(ClusterSpecification clusterSpecification)
+		throws ClusterDeploymentException {
+
+		final ClusterClient<String> clusterClient = deployClusterInternal(
+			KubernetesSessionClusterEntrypoint.class.getName(), clusterSpecification, false);
+
+		LOG.info("Create flink session cluster {} successfully, JobManager Web Interface: {}",
+			clusterId, clusterClient.getWebInterfaceURL());
+		return clusterClient;
+	}
+
+	@Override
+	public ClusterClient<String> deployJobCluster(ClusterSpecification clusterSpecification,
+			JobGraph jobGraph, boolean detached) throws ClusterDeploymentException {
+		throw new ClusterDeploymentException("Per job could not be supported now.");
+	}
+
+	private ClusterClient<String> deployClusterInternal(
+			String entryPoint,
+			ClusterSpecification clusterSpecification,
+			boolean detached) throws ClusterDeploymentException {
+
+		final ClusterEntrypoint.ExecutionMode executionMode = detached ?
+			ClusterEntrypoint.ExecutionMode.DETACHED
+			: ClusterEntrypoint.ExecutionMode.NORMAL;
+		flinkConfig.setString(ClusterEntrypoint.EXECUTION_MODE, executionMode.toString());
+
+		flinkConfig.setString(KubernetesConfigOptionsInternal.ENTRY_POINT_CLASS, entryPoint);
+
+		// Rpc(6123), blob(6124), rest(8081) taskManagerRpc(6122) port need to be exposed, so update them to fixed port.
+		flinkConfig.setString(BlobServerOptions.PORT, String.valueOf(Constants.BLOB_SERVER_PORT));
```

Comment:
So user setups on these `PORT`s have no power? It is a strong constraint?

Reply:
Nice catch.
Fixed port is required, because we need to expose the port in kubernetes. I will update the PR so that it could respect the user specified port value. If no `BlobServerOptions.PORT` and `TaskManagerOptions.RPC_PORT` are set, the default value will be used.